crates/allocator/src/bump.rs

```diff
@@ -0,0 +1,243 @@
+// Copyright 2018-2021 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A simple bump allocator.
+//!
+//! Its goal to have a much smaller footprint than the admittedly more full-featured `wee_alloc`
+//! allocator which is currently being used by ink! smart contracts.
+//!
+//! The heap which is used by this allocator is built from pages of Wasm memory (each page is `64KiB`).
+//! We will request new pages of memory as needed until we run out of memory, at which point we
+//! will crash with an `OOM` error instead of freeing any memory.
+
+use core::alloc::{
+    GlobalAlloc,
+    Layout,
+};
+
+/// A page in Wasm is `64KiB`
+const PAGE_SIZE: usize = 64 * 1024;
+
+static mut INNER: InnerAlloc = InnerAlloc::new();
+
+/// A bump allocator suitable for use in a Wasm environment.
+pub struct BumpAllocator;
+
+unsafe impl GlobalAlloc for BumpAllocator {
+    #[inline]
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        match INNER.alloc(layout) {
+            Some(start) => start as *mut u8,
+            None => core::ptr::null_mut(),
+        }
+    }
+
+    #[inline]
+    unsafe fn alloc_zeroed(&self, layout: Layout) -> *mut u8 {
+        // A new page in Wasm is guaranteed to already be zero initialized, so we can just use our
+        // regular `alloc` call here and save a bit of work.
+        self.alloc(layout)
+    }
+
+    #[inline]
+    unsafe fn dealloc(&self, _ptr: *mut u8, _layout: Layout) {}
+}
+
+#[cfg_attr(feature = "std", derive(Debug, Copy, Clone))]
+struct InnerAlloc {
+    /// Points to the start of the next available allocation.
+    next: usize,
+
+    /// The address of the upper limit of our heap.
+    upper_limit: usize,
+
+    /// The number of page requests made so far.
+    ///
+    /// This is meant to mimic the behavior exhibited by `core::arch::wasm32::memory_grow`
+    #[cfg(feature = "std")]
+    page_requests: usize,
```

Comment:
I would love if you could manage to get rid of this field. My feeling is that it is possible to use a shared field both for the testing and the production code. Isn't `page_requested = upper_limit / PAGE_SIZE` in the test case where the memory starts at `0`?